stable_baselines3/common/buffers.py

```diff
@@ -387,15 +380,81 @@ def get(self, batch_size: Optional[int] = None) -> Generator[RolloutBufferSample
+    def _get_samples(self, batch_inds: np.ndarray, env: Optional[VecNormalize] = None) -> RolloutBufferSamples:
+        data = (
+            self.observations[batch_inds],
+            self.actions[batch_inds],
+            self.values[batch_inds].flatten(),
+            self.log_probs[batch_inds].flatten(),
+            self.advantages[batch_inds].flatten(),
+            self.returns[batch_inds].flatten(),
+        )
         return RolloutBufferSamples(*tuple(map(self.to_torch, data)))
+
+
+class NstepReplayBuffer(ReplayBuffer):
+    def __init__(
+        self,
+        buffer_size: int,
+        observation_space: spaces.Space,
+        action_space: spaces.Space,
+        device: Union[th.device, str] = "cpu",
+        n_envs: int = 1,
+        optimize_memory_usage: bool = False,
+        n_step: int = 1,
+        gamma: float = 0.99,
+    ):
+        super().__init__(buffer_size, observation_space, action_space, device, n_envs, optimize_memory_usage)
+        self.n_step = int(n_step)
+        assert self.n_step <= buffer_size
+        self.gamma = float(gamma)
+        assert 0 <= gamma <= 1
+
+    def _get_samples(self, batch_inds: np.ndarray, env: Optional[VecNormalize] = None) -> ReplayBufferSamples:
+        assert not np.any(batch_inds == self.pos)
+        actions = self.actions[batch_inds, 0, :]
+
+        gamma = self.gamma
+
+        # Broadcasting turns 1dim arange matrix to 2 dimensional matrix that contains all
+        # the indices, % buffersize keeps us in buffer range
+        indices = (np.arange(self.n_step) + batch_inds.reshape(-1, 1)) % self.buffer_size
+
+        # two dim matrix of not dones. If done is true, then subsequent dones are turned to 0
+        # using accumulate. This ensures that we don't use invalid transitions
+        not_dones = np.multiply.accumulate(1 - self.dones[indices], 1).reshape(-1, self.n_step)
+
+        # vector of the discount factors
+        gammas = gamma ** np.arange(self.n_step)
+
+        # two dim matrix of rewards for the indices
+        rewards = self.rewards[indices].reshape(not_dones.shape)
+        rewards = self._normalize_reward(rewards, env)
+
+        # filter to ignore loops around the buffer, see:
+        # https://github.com/DLR-RM/stable-baselines3/pull/81#issuecomment-653741592
+        # basically we need to ignore indices that are equal to self.pos
+        # because the indice at self.pos is older
+        # not_dones that are from other transitions are filtered out.
+        filt = np.hstack([np.ones((len(batch_inds), 1)), not_dones[:, 1:]]) * (indices != self.pos)
+        filt = np.multiply.accumulate(filt, 1).reshape(filt.shape)
+
+        # weighted rewards
+        rewards = filt * gammas * rewards
+        rewards = np.add.reduce(rewards, 1).reshape(len(batch_inds), 1).astype(np.float32)
+
+        increments = np.add.reduce(filt, 1).astype(np.int).reshape(batch_inds.shape)
+        next_obs_indices = (increments + batch_inds - 1) % self.buffer_size
```

Comment:
I will stop here today. Overall looks ok with lot of nice numpy tricks ;)
I will try to do the rest next week.

my checking code:

```python
import numpy as np

buffer_size = 5
batch_inds = np.arange(5)
batch_inds[0] = 1
n_step = 2
gamma = 0.99
dones = np.zeros(5)
dones[3] = 1
rewards = np.ones(5) * 2
rewards[1] = 1
# Buffer pointer
pos = 0

print("rewards\n", rewards)
print("dones\n", dones)
print("buffer_pos=", pos)
print("batch_inds\n", batch_inds)

indices = np.arange(n_step) + batch_inds.reshape(-1, 1) 
indices = indices % buffer_size
print("indices\n", indices)

not_dones = np.multiply.accumulate(1 - dones[indices], 1).reshape(-1, n_step)
print("not_dones\n", not_dones)

gammas = gamma ** np.arange(n_step)
print("gammas\n", gammas)


rewards = rewards[indices].reshape(not_dones.shape)

print("rewards\n", rewards)

filt = np.hstack([np.ones((len(batch_inds), 1)), not_dones[:, 1:]]) * (indices != pos)

print("filter\n", filt)

filt = np.multiply.accumulate(filt, 1).reshape(filt.shape)

print("filter2\n", filt)

rewards = filt * gammas * rewards

print("weighted reward\n", rewards)

rewards = np.add.reduce(rewards, 1).reshape(len(batch_inds), 1).astype(np.float32)

print("reduced rewards\n", rewards)

increments = np.add.reduce(filt, 1).astype(np.int).reshape(batch_inds.shape)

print("increments\n", increments)

next_obs_indices = (increments + batch_inds - 1) % buffer_size

print("next_obs_indices\n", next_obs_indices)
```

Reply:
```
>>> import numpy as np
>>>
>>> buffer_size = 5
>>> batch_inds = np.arange(5)
>>> batch_inds[0] = 1
>>> n_step = 2
>>> gamma = 0.99
>>> dones = np.zeros(5)
>>> dones[3] = 1
>>> rewards = np.ones(5) * 2
>>> rewards[1] = 1
>>> # Buffer pointer
... pos = 0
>>>
>>> print("rewards\n", rewards)
rewards
 [2. 1. 2. 2. 2.]
>>> print("dones\n", dones)
dones
 [0. 0. 0. 1. 0.]
>>> print("buffer_pos=", pos)
buffer_pos= 0
>>> print("batch_inds\n", batch_inds)
batch_inds
 [1 1 2 3 4]
>>> indices = np.arange(n_step) + batch_inds.reshape(-1, 1)
>>> indices = indices % buffer_size
>>> print("indices\n", indices)
indices
 [[1 2]
 [1 2]
 [2 3]
 [3 4]
 [4 0]]
```
We can agree that all of these are correct :)
So `not_dones` performs a forward filtering on the future states to consider.
It is expected that when the current transition is terminal, the whole row will be 0.
This is accounted for later.
```
>>> not_dones = np.multiply.accumulate(1 - dones[indices], 1).reshape(-1, n_step)
>>> print("not_dones\n", not_dones)
not_dones
 [[1. 1.]
 [1. 1.]
 [1. 0.]
 [0. 0.]
 [1. 1.]]
```
As  expected the 4th row is 0.

```
>>> gammas = gamma ** np.arange(n_step)
>>> print("gammas\n", gammas)
gammas
 [1.   0.99]
>>> rewards = rewards[indices].reshape(not_dones.shape)
>>> print("rewards\n", rewards)
rewards
 [[1. 2.]
 [1. 2.]
 [2. 2.]
 [2. 2.]
 [2. 2.]]
```
Nothing to add above.

This is the interesting part.
The first thing we do here is to correct the `not_dones` in the first column to be equal to 1, because this is the immediate reward. Because of `multiply.accumulate`, `not_dones` will have 0s for transitions from other episodes.
```
>>> filt = np.hstack([np.ones((len(batch_inds), 1)), not_dones[:, 1:]]) * (indices != pos)
>>> print("filter\n", filt)
filter
 [[1. 1.]
 [1. 1.]
 [1. 0.]
 [1. 0.]
 [1. 0.]]
>>> filt = np.multiply.accumulate(filt, 1).reshape(filt.shape)
>>>
>>> print("filter2\n", filt)
filter2
 [[1. 1.]
 [1. 1.]
 [1. 0.]
 [1. 0.]
 [1. 0.]]
>>>
```
There is a bug here, third row, second column should be one. I will fix it and add it to the test cases.
Last row is correct, `pos==0`, so we shouldn't use that transition.

```
>>> rewards = filt * gammas * rewards
>>>
>>> print("weighted reward\n", rewards)
weighted reward
 [[1.   1.98]
 [1.   1.98]
 [2.   0.  ]
 [2.   0.  ]
 [2.   0.  ]]
>>>
>>> rewards = np.add.reduce(rewards, 1).reshape(len(batch_inds), 1).astype(np.float32)
>>>
>>> print("reduced rewards\n", rewards)
reduced rewards
 [[2.98]
 [2.98]
 [2.  ]
 [2.  ]
 [2.  ]]
```
Yup, third row should have had a reward of 3.98.
```
>>>
>>> increments = np.add.reduce(filt, 1).astype(np.int).reshape(batch_inds.shape)
>>>
>>> print("increments\n", increments)
increments
 [2 2 1 1 1]
```
Same bug as above, increments[2] should have been 2.
```
>>>
>>> next_obs_indices = (increments + batch_inds - 1) % buffer_size
>>>
>>> print("next_obs_indices\n", next_obs_indices)
next_obs_indices
 [2 2 2 3 4]
```
Same bug as above.

Reply:
This is solved now.